File: M1files/parser.py
from typing import (
    Optional,
    List,
    Dict,
    Callable
    )
from enum import IntEnum
from M1files.ast import (
    Block,
    Boolean,
    Call,
    Expression,
    ExpressionStatement,
    Function,
    Identifier,
    If,
    Infix,
    LetStatement,
    Prefix,
    Program,
    Statement,
    ReturnStatement
    )
from M1files.lexer import Lexer
from M1files.tokens import (
    Token,
    TokenType
)

PrefixParseFn=Callable[[],Optional[Expression]]
InfixParseFn=Callable[[Expression],Optional[Expression]]
PrefixParseFns:Dict[TokenType,PrefixParseFn]#={}
InfixParseFns:Dict[TokenType,InfixParseFn]#={}
class Precendece(IntEnum):
    LOWEST = 1
    EQUALS = 2
    LESSGREATER = 3
    SUM = 4
    PRODUCT = 5
    PREFIX = 6
    CALL = 7

PRECEDENTES:Dict[TokenType,Precendece]={
    TokenType.EQ:Precendece.EQUALS,
    TokenType.LT:Precendece.LESSGREATER,
    TokenType.GT:Precendece.LESSGREATER,
    TokenType.DIF:Precendece.EQUALS,
    TokenType.PLUS:Precendece.SUM,
    TokenType.MINUS:Precendece.SUM,
    TokenType.MULTIPLICATION:Precendece.PRODUCT,
    TokenType.DIVISION:Precendece.PRODUCT,
    TokenType.LBRACE:Precendece.CALL,

}

class Parser:
    def __init__(self, lexer:Lexer)->None:
        self._lexer = lexer
        self._current_token:Optional[Token] = None
        self._peek_token:Optional[Token] = None
        self._advance_tokens()
        self._advance_tokens()
    def parse_program(self)->Program:
        program:Program=Program(statements=[])
        assert self._current_token is not None
        while self._current_token.type != TokenType.EOF:
            statement = self._parse_statement()
            if statement is not None:
                program.statements.append(statement)
            self._advance_tokens()
        return program
    def _advance_tokens(self)->None:
        self._current_token=self._peek_token
        self._peek_token=self._lexer.next_token()
    def _expected_token(self, token_type:TokenType) -> bool:
        assert self._peek_token is not None
        if self._peek_token.type == token_type:
            self._advance_tokens()
            return True
        return False
    def _parse_let_statement(self)->Optional[LetStatement]:
        assert self._current_token is not None
        let_statement:LetStatement = LetStatement(token = self._current_token)
        if not self._expected_token(TokenType.IDENT):
            return None
        let_statement.name = Identifier(token = self._current_token, value = self._current_token.literal)
        if not self._expected_token(TokenType.ASSIGN):
            return None
        # TODO Falta evaluar la expresión
        while self._current_token.type != TokenType.SEMICOLON:
            self._advance_tokens() 
        
        return let_statement

    def _parse_statement(self)->Optional[Statement]:
        assert self._current_token is not None
        if self._current_token.type == TokenType.LET:
            return self._parse_let_statement()
        elif self._current_token.type == TokenType.RETURN:
            return self._parse_return_statement()
        else:
            return None
        
    def _parse_block(self)-> Block:
        assert self._current_token is not None
        block_statement:Block = Block(token = self._current_token, statements = [])
        self._advance_tokens()
        while not self._current_token.type == TokenType.RBRACE \
        and not self._current_token.type == TokenType.EOF:
            statement = self._parse_statement()
            if statement:
                block_statement.statements.append(statement)
            self._advance_tokens() 
        return block_statement
    def _parse_if(self) -> Optional[If]:
        assert self._current_token is not None
        if_expression:If = If(token = self._current_token)

        if not self._expected_token(TokenType.LPAREN):
            return None
        self._advance_tokens()

        if_expression.condition = self._parse_expression(Precendece.LOWEST)

        if not self._expected_token(TokenType.RPAREN):
            return None
        if not self._expected_token(TokenType.LBRACE):
            return None
        if_expression.consequence = self._parse_block()

        assert self._peek_token is not None
        if self._peek_token.type == TokenType.ELSE:
            self._advance_tokens()
            if not self._expected_token(TokenType.LBRACE):
                return None
            if_expression.alternative = self._parse_block()

        return if_expression
    
    def _parse_expression(self, precender:Precendece)->Optional[Expression]:
        assert self._current_token is not None
        try:
            prefix_parse_fn:self._prefix_parse_fns[self._current_token.type]
        except KeyError:
            message = f'No se encontró una función para parsear el token {self._current_token.type}'
            self._errors.append(message)
            return None

File: M1files/Parser.java
package M1files;

import java.util.List;
//import java.util.Optional;
import java.util.Map;
import java.util.ArrayList;
import java.util.Collections;
import java.util.EnumMap;

// Import classes from M1files package
/*import M1files.Token;
import M1files.TokenType;
import M1files.Lexer;
import M1files.Block;
import M1files.Boolean;
import M1files.Call;
import M1files.Expression;
import M1files.ExpressionStatement;
import M1files.Function;
import M1files.Identifier;
import M1files.If;
import M1files.Infix;
import M1files.Integer;
import M1files.LetStatement;
import M1files.Prefix;
import M1files.Program;
import M1files.ReturnStatement;
import M1files.Statement;*/

public class Parser {
    private Lexer lexer;
    private Token currentToken;
    private Token peekToken;
    private List<String> errors;
    
    enum Precedence {
        LOWEST,
        EQUALS, 
        LESSGREATER, 
        SUM, 
        PRODUCT, 
        PREFIX, 
        CALL
    }

    interface PrefixParseFn extends java.util.function.Function<Void, Expression> {}
    interface InfixParseFn extends java.util.function.Function<Expression, Expression> {}
    Map<TokenType, PrefixParseFn> prefixParseFns = new EnumMap<>(TokenType.class);
    Map<TokenType, InfixParseFn> infixParseFns = new EnumMap<>(TokenType.class);
    Map<TokenType, Precedence> precedences = new EnumMap<>(TokenType.class);
    
    public Parser(Lexer lexer){
        precedences.put(TokenType.EQ, Precedence.EQUALS);
        precedences.put(TokenType.DIF, Precedence.EQUALS);
        precedences.put(TokenType.LT, Precedence.LESSGREATER);
        precedences.put(TokenType.GT, Precedence.LESSGREATER);
        precedences.put(TokenType.PLUS, Precedence.SUM);
        precedences.put(TokenType.MINUS, Precedence.SUM);
        precedences.put(TokenType.DIVISION, Precedence.PRODUCT);
        precedences.put(TokenType.MULTIPLICATION, Precedence.PRODUCT);
        precedences.put(TokenType.LBRACE, Precedence.CALL);

        this.lexer = lexer;
        this.currentToken = null;
        this.peekToken = null;
        this.errors = new ArrayList<>();

        this.prefixParseFns = registerPrefixParseFns();
        this.infixParseFns = registerInfixParseFns();

        advanceTokens();
        advanceTokens();
    }
    
    public Program parseProgram() {
        Program program = new Program(new ArrayList<>());
        assert currentToken != null;
        while (currentToken.getTokenType() != TokenType.EOF) {
            Statement statement = parseStatement();
            if (statement != null)
                program.addStatement(statement);
            advanceTokens();
        }
        return program;
    }
    
    private void advanceTokens() {
        currentToken = peekToken;
        peekToken = lexer.nextToken();
    }

    private Precedence currentPrecedence() {
        assert currentToken != null;
        try {
            return precedences.get(currentToken.getTokenType());
        } catch (NullPointerException e) {
            return Precedence.LOWEST;
        }
    }
    
    private boolean expectedToken(TokenType tokenType) {
        assert peekToken != null;
        if (peekToken.getTokenType() == tokenType) {
            advanceTokens();
            return true;
        }
        expectedTokenError(tokenType);
        return false;
    }

    private void expectedTokenError(TokenType tokenType) {
        assert peekToken != null;
        String message = "A " + tokenType + " token was expected, but " + peekToken.getTokenType() + " was found";
        errors.add(message);
    }
    
    private Block parseBlock() {
        assert currentToken != null;
        Block blockStatement = new Block(currentToken, new ArrayList<>());
        advanceTokens();
        while (currentToken.getTokenType() != TokenType.RBRACE && currentToken.getTokenType() != TokenType.EOF) {
            Statement statement = parseStatement();
            if (statement != null) {
                blockStatement.addStatement(statement);
            }
            advanceTokens();
        }
        return blockStatement;
    }

    private Boolean parseBoolean() {
        assert currentToken != null;
        return new Boolean(currentToken, currentToken.getTokenType() == TokenType.TRUE);
    }
    
    private Call parseCall(Expression function) {
        assert currentToken != null;
        Call call = new Call(currentToken, function, new ArrayList<>());
        call.setArguments(parseCallArguments());
        return call;
    }

    private List<Expression> parseCallArguments() {
        List<Expression> arguments = new ArrayList<>();
        assert peekToken != null;
        if (peekToken.getTokenType() == TokenType.RPAREN) {
            advanceTokens();
            return arguments;
        }
        advanceTokens();
        Expression expression = parseExpression(Precedence.LOWEST);
        if (expression != null) {
            arguments.add(expression);
        }
        while (peekToken.getTokenType() == TokenType.COMMA) {
            advanceTokens();
            advanceTokens();
            expression = parseExpression(Precedence.LOWEST);
            if (expression != null) {
                arguments.add(expression);
            }
        }
        if (!expectedToken(TokenType.RPAREN)) {
            return null;
        }
        return arguments;
    }

    private Expression parseExpression(Precedence precedence) {
        assert currentToken != null;
        try {
            PrefixParseFn prefixParseFn = prefixParseFns.get(currentToken.getTokenType());
            if (prefixParseFn == null) {
                String message = "Function to parse was not found for token type: " + currentToken.getTokenType();
                errors.add(message);
                return null;
            }
            Expression leftExpression = prefixParseFn.apply(null);
            assert peekToken != null;
            while (!peekToken.getTokenType().equals(TokenType.SEMICOLON) && precedence.compareTo(peekPrecedence()) < 0) {
                try {
                    InfixParseFn infixParseFn = infixParseFns.get(peekToken.getTokenType());
                    advanceTokens();
                    assert leftExpression != null;
                    leftExpression = infixParseFn.apply(leftExpression);
                } catch (Exception e) {
                    return leftExpression;
                }
            }
            return leftExpression;
        } catch (Exception e) {
            return null;
        }
    }

    private ExpressionStatement parseExpressionStatement() {
        assert currentToken != null;
        ExpressionStatement expressionStatement = new ExpressionStatement(currentToken, null);
        expressionStatement.setExpression(parseExpression(Precedence.LOWEST));
        assert peekToken != null;
        if (peekToken.getTokenType() == TokenType.SEMICOLON) {
            advanceTokens();
        }
        return expressionStatement;
    }
    
    private Expression parseGroupedExpression() {
        advanceTokens();
        Expression expression = parseExpression(Precedence.LOWEST);
        if (!expectedToken(TokenType.RPAREN)) {
            return null;
        }
        return expression;
    }

    private Function parseFunction() {
        assert currentToken != null;
        Function function = new Function(currentToken, new ArrayList<>(), null);
        if (!expectedToken(TokenType.LPAREN)) {
            return null;
        }
        function.setParameters(parseFunctionParameters());
        if (!expectedToken(TokenType.LBRACE)) {
            return null;
        }
        function.setBody(parseBlock());
        return function;
    }

    private List<Identifier> parseFunctionParameters() {
        List<Identifier> params = new ArrayList<>();
        assert peekToken != null;
        if (peekToken.getTokenType() == TokenType.RPAREN) {
            advanceTokens();
            return params;
        }
        advanceTokens();
        assert currentToken != null;
        Identifier identifier = new Identifier(currentToken);
        params.add(identifier);
        while (peekToken.getTokenType() == TokenType.COMMA) {
            advanceTokens();
            advanceTokens();
            assert currentToken != null;
            identifier = new Identifier(currentToken);
            params.add(identifier);
        }
        if (!expectedToken(TokenType.RPAREN)) {
            return Collections.emptyList();
        }
        return params;
    }

    private Identifier parseIdentifier() {
        assert currentToken != null;
        return new Identifier(currentToken);
    }
    
    private If parseIf() {
        assert currentToken != null;
        If ifExpression = new If(currentToken, null, null, null);
        if (!expectedToken(TokenType.LPAREN)) {
            return null;
        }
        advanceTokens();
        ifExpression.setCondition(parseExpression(Precedence.LOWEST));
        if (!expectedToken(TokenType.RPAREN)) {
            return null;
        }
        if (!expectedToken(TokenType.LBRACE)) {
            return null;
        }
        ifExpression.setConsequence(parseBlock());
        assert peekToken != null;
        if (peekToken.getTokenType() == TokenType.ELSE) {
            advanceTokens();
            if (!expectedToken(TokenType.LBRACE)) {
                return null;
            }
            ifExpression.setAlternative(parseBlock());
        }
        return ifExpression;
    }

    private Infix parseInfixExpression(Expression left) {
        assert currentToken != null;
        Infix infix = new Infix(currentToken, left, currentToken.getLiteral(), null);
        Precedence precedence = currentPrecedence();
        advanceTokens();
        infix.setRight(parseExpression(precedence));
        return infix;
    }
    
    private Integer parseInteger() {
        assert currentToken != null;
        try {
            return new Integer(currentToken, java.lang.Integer.parseInt(currentToken.getLiteral()));
            //integer.setValue(Integer(currentToken, currentToken.getLiteral()));
        } catch (NumberFormatException e) {
            String message = "Could not parse " + currentToken.getLiteral() + " as integer";
            errors.add(message);
            return null;
        }
    }
    
    private LetStatement parseLetStatement() {
        assert currentToken != null;
        LetStatement letStatement = new LetStatement(currentToken, null, null);
        if (!expectedToken(TokenType.IDENTIFIER)) {
            return null;
        }
        letStatement.setName(parseIdentifier());
        if (!expectedToken(TokenType.ASSIGN)) {
            return null;
        }
        advanceTokens();
        letStatement.setValue(parseExpression(Precedence.LOWEST));
        assert peekToken != null;
        if (peekToken.getTokenType() == TokenType.SEMICOLON) {
            advanceTokens();
        }
        return letStatement;
    }
    
    private Prefix parsePrefixExpression() {
        assert currentToken != null;
        Prefix prefix = new Prefix(currentToken, currentToken.getLiteral(), null);
        advanceTokens();
        prefix.setRight(parseExpression(Precedence.PREFIX));
        return prefix;
    }

    private ReturnStatement parseReturnStatement() {
        assert currentToken != null;
        ReturnStatement returnStatement = new ReturnStatement(currentToken, null);
        advanceTokens();
        returnStatement.setReturnValue(parseExpression(Precedence.LOWEST));
        assert peekToken != null;
        if (peekToken.getTokenType() == TokenType.SEMICOLON) {
            advanceTokens();
        }
        return returnStatement;
    }
    
    private Statement parseStatement() {
        assert currentToken != null;
        if (currentToken.getTokenType() == TokenType.LET) {
            return parseLetStatement();
        } else if (currentToken.getTokenType() == TokenType.RETURN) {
            return parseReturnStatement();
        } else {
            return parseExpressionStatement();
        }
    }

    private Precedence peekPrecedence() {
        assert peekToken != null;
        try {
            return precedences.get(peekToken.getTokenType());
        } catch (NullPointerException | IllegalArgumentException e) {
            return Precedence.LOWEST;
        }
    }

    private Map<TokenType, InfixParseFn> registerInfixParseFns() {
        Map<TokenType, InfixParseFn> infixFns = new EnumMap<>(TokenType.class);
        infixFns.put(TokenType.PLUS, this::parseInfixExpression);
        infixFns.put(TokenType.MINUS, this::parseInfixExpression);
        infixFns.put(TokenType.DIVISION, this::parseInfixExpression);
        infixFns.put(TokenType.MULTIPLICATION, this::parseInfixExpression);
        infixFns.put(TokenType.EQ, this::parseInfixExpression);
        infixFns.put(TokenType.DIF, this::parseInfixExpression);
        infixFns.put(TokenType.LT, this::parseInfixExpression);
        infixFns.put(TokenType.GT, this::parseInfixExpression);
        infixFns.put(TokenType.LPAREN, this::parseCall);
        return infixFns;
    }
    
    private Map<TokenType, PrefixParseFn> registerPrefixParseFns() {
        Map<TokenType, PrefixParseFn> prefixFns = new EnumMap<>(TokenType.class);
        prefixFns.put(TokenType.FALSE, token -> parseBoolean());
        prefixFns.put(TokenType.FUNCTION, token -> parseFunction());
        prefixFns.put(TokenType.IDENTIFIER, token -> parseIdentifier());
        prefixFns.put(TokenType.IF, token -> parseIf());
        prefixFns.put(TokenType.INTEGER, token -> parseInteger());
        prefixFns.put(TokenType.LPAREN, token -> parseGroupedExpression());
        prefixFns.put(TokenType.MINUS, token -> parsePrefixExpression());
        prefixFns.put(TokenType.NEGATION, token -> parsePrefixExpression());
        prefixFns.put(TokenType.TRUE, token -> parseBoolean());
        return prefixFns;
        }
}